src/auth.rs

```diff
@@ -0,0 +1,115 @@
+use cached::proc_macro::cached;
+use serde::{Deserialize, Serialize};
+
+use crate::{ApiEnvironment, ApiError, Mpesa, MpesaError, MpesaResult};
+
+const AUTHENTICATION_URL: &str = "/oauth/v1/generate?grant_type=client_credentials";
+
+/// Response returned from the authentication function
+#[derive(Debug, Serialize, Deserialize)]
+pub struct AuthenticationResponse {
+    /// Access token which is used as the Bearer-Auth-Token
+    pub access_token: String,
+    /// Expiry time in seconds
+    pub expiry_in: u64,
+}
+
+impl std::fmt::Display for AuthenticationResponse {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "token :{} expires in: {}",
+            self.access_token, self.expiry_in
+        )
+    }
+}
```

Comment:
Minor: code guideline for the code base (should probably be documented) follows "the bigger picture" rule, therefore when someone opens this file for the first time, they should see the most important code first, which in this case is the `auth` function.

I suggest you move the highlighted code below this function